File: 3_Structural_Design_Patterns/3.04_Facade/Java_Snippet/Facade.java
class Television {
    public void on() {
        System.out.println("Turning on TV");
    }

    public void off() {
        System.out.println("Turning off TV");
    }
}

class SoundSystem {
    public void on() {
        System.out.println("Turning on SoundSystem");
    }

    public void off() {
        System.out.println("Turning off SoundSystem");
    }
}

class DVDPlayer {
    public void on() {
        System.out.println("Turning on DVDPlayer");
    }

    public void off() {
        System.out.println("Turning off DVDPlayer");
    }

    public void play() {
        System.out.println("Playing DVD");
    }
}

class HomeTheatreFacade {
    private Television tv;
    private SoundSystem soundSystem;
    private DVDPlayer dvdPlayer;

    public HomeTheatreFacade(Television tv, SoundSystem soundSystem, DVDPlayer dvdPlayer) {
        this.tv = tv;
        this.soundSystem = soundSystem;
        this.dvdPlayer = dvdPlayer;
    }

    public void watchMovie(String movie) {
        System.out.println("Get ready to watch a movie...");
        tv.on();
        soundSystem.on();
        dvdPlayer.on();
        dvdPlayer.play();
    }

    public void endMovie() {
        System.out.println("Shutting down the home theater...");
        tv.off();
        soundSystem.off();
        dvdPlayer.off();
    }
}

public class Facade {
    public static void main(String[] args) {
        Television tv = new Television();
        SoundSystem soundSystem = new SoundSystem();
        DVDPlayer dvdPlayer = new DVDPlayer();

        HomeTheatreFacade homeTheater = new HomeTheatreFacade(tv, soundSystem, dvdPlayer);

        homeTheater.watchMovie("Inception");
        System.out.println("\nMovie ends, shutting down the home theater...");
        homeTheater.endMovie();
    }
}

File: 3_Structural_Design_Patterns/3.04_Facade/Python_Snippet/Facade.py
class Television:
    def on(self):
        print("Turning on TV")

    def off(self):
        print("Turning off TV")


class SoundSystem:
    def on(self):
        print("Turning on SoundSystem")

    def off(self):
        print("Turning off SoundSystem")


class DVDPlayer:
    def on(self):
        print("Turning on DVDPlayer")

    def off(self):
        print("Turning off DVDPlayer")

    def play(self):
        print("Playing DVD")


class HomeTheatreFacade:
    def __init__(self, tv, sound_system, dvd_player):
        self.tv = tv
        self.sound_system = sound_system
        self.dvd_player = dvd_player

    def watch_movie(self, movie):
        print("Get ready to watch a movie...")
        self.tv.on()
        self.sound_system.on()
        self.dvd_player.on()
        self.dvd_player.play()

    def end_movie(self):
        print("Shutting down the home theater...")
        self.tv.off()
        self.sound_system.off()
        self.dvd_player.off()


if __name__ == "__main__":
    tv = Television()
    sound_system = SoundSystem()
    dvd_player = DVDPlayer()

    home_theater = HomeTheatreFacade(tv, sound_system, dvd_player)

    home_theater.watch_movie("Inception")
    print("\nMovie ends, shutting down the home theater...")
    home_theater.end_movie()